pgx/src/datum/numeric.rs

```diff
@@ -7,218 +7,225 @@ All rights reserved.
-                        // and don't leak the 'inet' datum Postgres created
-                        pg_sys::pfree(datum.cast_mut_ptr());
-                    }
-
-                    // we have it as a valid String
-                    Ok(Numeric(v.clone()))
-                })
-                .catch_when(PgSqlErrorCode::ERRCODE_INVALID_TEXT_REPRESENTATION, |_| {
-                    Err(Error::custom(format!("invalid Numeric value: {}", v)))
-                })
-                .execute()
+impl Drop for AnyNumeric {
+    fn drop(&mut self) {
+        if self.need_pfree {
+            unsafe {
+                pg_sys::pfree(self.inner.cast());
             }
         }
-
-        deserializer.deserialize_any(NumericVisitor)
     }
 }
 
-impl From<i8> for Numeric {
-    fn from(val: i8) -> Self {
-        Numeric(val.to_string())
+impl Display for AnyNumeric {
+    fn fmt(&self, fmt: &mut Formatter<'_>) -> fmt::Result {
+        let numeric_out = unsafe {
+            direct_function_call::<&CStr>(pg_sys::numeric_out, vec![self.as_datum()]).unwrap()
+        };
+        let s = numeric_out.to_str().expect("numeric_out is not a valid UTF8 string");
+        fmt.write_str(s)
     }
 }
 
-impl From<i16> for Numeric {
-    fn from(val: i16) -> Self {
-        Numeric(val.to_string())
+impl Debug for AnyNumeric {
+    fn fmt(&self, fmt: &mut Formatter<'_>) -> fmt::Result {
+        write!(fmt, "AnyNumeric(inner: {self}, need_pfree:{})", self.need_pfree)
     }
 }
 
-impl From<i32> for Numeric {
-    fn from(val: i32) -> Self {
-        Numeric(val.to_string())
+impl<const P: u32, const S: u32> Display for Numeric<P, S> {
+    fn fmt(&self, fmt: &mut Formatter<'_>) -> Result<(), fmt::Error> {
+        write!(fmt, "{}", self.0)
     }
 }
 
-impl From<i64> for Numeric {
-    fn from(val: i64) -> Self {
-        Numeric(val.to_string())
+#[inline(always)]
+pub(crate) const fn make_typmod(precision: u32, scale: u32) -> i32 {
+    let (precision, scale) = (precision as i32, scale as i32);
+    match (precision, scale) {
+        (0, 0) => -1,
+        (p, s) => ((p << 16) | s) + pg_sys::VARHDRSZ as i32,
     }
 }
 
-impl From<u8> for Numeric {
-    fn from(val: u8) -> Self {
-        Numeric(val.to_string())
-    }
+/// Which way does a [`AnyNumeric`] sign face?
+pub enum Sign {
+    Negative,
+    Positive,
+    Zero,
 }
 
-impl From<u16> for Numeric {
-    fn from(val: u16) -> Self {
-        Numeric(val.to_string())
+impl AnyNumeric {
+    /// Returns the sign of this [`AnyNumeric`]
+    pub fn sign(&self) -> Sign {
+        unsafe {
+            let sign: AnyNumeric =
+                direct_function_call(pg_sys::numeric_sign, vec![self.as_datum()]).unwrap();
+            let zero: AnyNumeric = 0.try_into().unwrap();
+            if sign < zero {
+                return Sign::Negative;
+            } else if sign > zero {
+                return Sign::Positive;
+            } else {
+                return Sign::Zero;
+            }
```

Comment:
I think we can just do something like this (with NaN handling added) and remove the need for a call into Postgres?
```suggestion
            let zero: AnyNumeric = 0.try_into().unwrap();
            if self < zero {
                return Sign::Negative;
            } else if self > zero {
                return Sign::Positive;
            } else {
                return Sign::Zero;
            }
```

Reply:
I'm not sure.  `numeric_sign()` has some handling for -/+Infinity too.  

I'll definitely fix this to account for NaN.